orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -62,192 +62,220 @@ IMPLEMENT_GRADIENT_BUILDER(GetMatMulGradient) {
+
+          std::vector<int64_t> dY_shape_2d{-1, N};
+          NodeDef dY_shape_2d_node = ConstantValueNode(dY_shape_2d, Name("dY_shape_2d"));
+          ArgDef dY_shape_2d_arg = dY_shape_2d_node.output_args[0];
+          result.push_back(dY_shape_2d_node);
+
+          NodeDef zero_constant_node = ZeroConstantNode();
+          ArgDef ZERO = zero_constant_node.output_args[0];
+          result.push_back(zero_constant_node);
+
+          result.push_back(
+              NodeDef("Reshape",
+                      {A, A_shape_2d_arg},
+                      {IA("A_reshape_2d")}));
+          result.push_back(
+              NodeDef("Reshape",
+                      {GO(0), dY_shape_2d_arg},
+                      {IA("dY_reshape_2d")}));
+
+          // dB = A' * dY
+          std::vector<AttributeProto> attrs(shared_attributes);
+          attrs.push_back(transpose_first_input);
+          result.push_back(
+              NodeDef("Gemm",
+                      {IA("A_reshape_2d"), IA("dY_reshape_2d"), ZERO},
+                      {GI(1)},
+                      attrs));
+        } else {
+          int64_t A_rank = A_shape.size();
+          std::vector<int64_t> A_perm(A_rank);
+          std::iota(A_perm.begin(), A_perm.end(), 0);
+          std::swap(A_perm[A_rank - 1], A_perm[A_rank - 2]);
+
+          std::vector<Dimension> output_shape;
+          for (size_t i = 0; i < Y_shape.size() - 2; i++) {
+            output_shape.push_back(Y_shape[i]);
+          }
+          output_shape.push_back(A_shape[A_shape.size() - 1]);
+          output_shape.push_back(Y_shape[Y_shape.size() - 1]);
+
+          std::vector<int64_t> B_axes;
+          ComputeBroadcastBackwardAxes(B_shape, output_shape, &B_axes, nullptr);
+
+          result.push_back(
+              NodeDef("Transpose",
+                      {A},
+                      {IA("A_t")},
+                      {MakeAttribute("perm", A_perm)}));
+
+          ArgDef matmul_out = B_axes.size() > 0 ? IA("PreReduceGrad1") : GI(1);
+
+          result.push_back(
+              NodeDef("MatMul",
+                      {IA("A_t"), GO(0)},
+                      {matmul_out}));
+
+          if (B_axes.size() > 0) {
+            result.push_back(
+                NodeDef("ReduceSum",
+                        {IA("PreReduceGrad1")},
+                        {IA("ReduceGrad1")},
+                        {{"keepdims", MakeAttribute("keepdims", int64_t(0))},
+                         {"axes", MakeAttribute("axes", B_axes)}}));
+            result.push_back(
+                NodeDef("Shape",
+                        {B},
+                        {IA("B_shape")}));
+            result.push_back(
+                NodeDef("Reshape",
+                        {IA("ReduceGrad1"), IA("B_shape")},
+                        {GI(1)}));
+          }
         }
       }
     }
-  } else {
-    ORT_THROW("Matmul Gradient Builder shouldn't reach here. ");
+  } catch (const onnxruntime::OnnxRuntimeShapeException& e) {
```

Comment:
>const onnxruntime::OnnxRuntimeShapeException& e [](start = 11, length = 47)

make it onnxruntime::OnnxRuntimeShapeException e ?
windows would complain e is not used